polkadot-parachains/parachains-common/src/impls.rs

```diff
@@ -281,4 +322,59 @@ mod tests {
 			"AssetsFrom should allow assets from any of its interior locations"
 		);
 	}
+
+	#[test]
+	fn assets_from_convert_correctly() {
+		parameter_types! {
+			pub const Local: MultiLocation = Here.into();
+			pub Remote: MultiLocation = MultiLocation::new(1, X1(Parachain(1000)));
+		}
+		let local_asset_location = Local::get().pushed_with_interior(GeneralIndex(42)).unwrap();
+		let remote_asset_location = Remote::get().pushed_with_interior(GeneralIndex(42)).unwrap();
+		let mut asset_id =
+			AsPrefixedGeneralIndexFromLocalOrRemote::<Local, Remote, u32, JustTry>::convert_ref(
+				&local_asset_location,
+			)
+			.unwrap_or(u32::default());
+		assert_eq!(asset_id, 42);
+		asset_id =
+			AsPrefixedGeneralIndexFromLocalOrRemote::<Local, Remote, u32, JustTry>::convert_ref(
+				&remote_asset_location,
+			)
+			.unwrap_or(u32::default());
+		assert_eq!(asset_id, 42);
```

Comment:
Previously I have tried to summarize your problem as the following:

> if I'm hearing you correctly, you're saying that the `GeneralIndex` is used to specify which one of the many assets in statemine that you want to transfer to the destination chain via reserve transfers? In other words, your flow is exactly the same as `InititateReserveWithdraw`, but with the ability to specify an asset that is not KSM, correct?

But this unit test seems to indicate that you're doing something more -- you're trying to represent a remote asset containing a `GeneralIndex` locally as a `GeneralIndex` as well, e.g. a `../Parachain(1000)/GeneralIndex(42)` gets converted and represented as `./GeneralIndex(42)`. Is this what you are trying to do?

If so, then this is absolutely not the right thing to do. Assets should always have __ONE__ canonical `MultiLocation` to represent it, and as such, an asset located at `../Parachain(1000)/GeneralIndex(42)` __MUST__ be interpreted as a different asset from an asset located at `./GeneralIndex(42)`. If the intention is to mint a derivative asset of the one located at `../Parachain(1000)/GeneralIndex(42)`, then it is the `AssetTransactor` logic of your chain that needs to change, and not the XCM configuration.

Reply:
Pondering a bit more into this, I think what you are trying to solve is the reanchoring issue, in which case, this is an operation that should have been done by the XCM sender, and not the recipient, i.e. when sending an asset containing a `MultiLocation` of `../Parachain(1000)/GeneralIndex(42)` to `../Parachain(1000)`, the XCM executor [should have reanchored this `AssetId` to `./GeneralIndex(42)`](https://github.com/paritytech/polkadot/blob/2d84ad65fb3ac54acab7486426267b3ea5b72681/xcm/xcm-executor/src/lib.rs#L312). Did this not happen when you're testing?

Reply:
@KiChjang @apopiak I do not think we can achive that without hack in cumulus, let us put more details here:
  
- transfer asset from statemine to our chain with `TransferReserveAsset`

since [`Local`](https://github.com/paritytech/cumulus/blob/e70f7fd0162fef6ca63bc31466bb2f66469310fa/polkadot-parachains/westmint/src/lib.rs#L467) defined here original asset sent from statemine should be represented as `./GeneralIndex(42)`  and after the [reanchor logic]( https://github.com/paritytech/polkadot/blob/2d84ad65fb3ac54acab7486426267b3ea5b72681/xcm/xcm-executor/src/lib.rs#L312) you mentioned above asset will be converted to `../Parachain(1000)/GeneralIndex(42)` and in our chain we should add  some [adaptor logic](https://github.com/bifrost-finance/bifrost/blob/c7eb1bef7abfdeb6562aeda29ab76805308b77f5/runtime/common/src/xcm_impl.rs#L153-L158) to handle this

- transfer asset from our chain to statemine with `InitiateReserveWithdraw`

 I agree `Assets should always have ONE canonical MultiLocation to represent` so original asset sent from our chain should still be `../Parachain(1000)/GeneralIndex(42)` but after the [reanchor logic here](https://github.com/paritytech/polkadot/blob/24758c5ab5e6c7f99e9af08e583f6084f5d087a6/xcm/xcm-executor/src/lib.rs#L410) will still be `../Parachain(1000)/GeneralIndex(42)` and we can not convert it back to `./GeneralIndex(42)` so statemine will not understand without hack

Reply:
A couple of observations:

1. What you interact on your local chain should _always_ be `./GeneralIndex(42)` and not `../Parachain(1000)/GeneralIndex(42)`, i.e. you interact with the derivative asset and never the reserve asset. Therefore when you do a reserve asset transfer, you will first execute a `WithdrawAsset(GeneralIndex(42))` and you will get some `GeneralIndex(42)` stored in your chain's holding register.
2. The next step in the process is `InitiateReserveWithdraw`, which takes the derivative asset in the holding register and sends it back to Statemine. Before that happens though, what should happen on your location chain is that it should burn and convert the derivative asset (i.e. `GeneralIndex(42)`) to the reserve asset (i.e. `../Parachain(1000)/GeneralIndex(42)`).
3. The receiver should never be responsible for converting between derivative asset IDs and reserve asset IDs. Doing so would mean that all reserve assets should "register" their derivatives with the reserve chain, which is absolutely not the direction we want to head towards.

I've looked into the code for the XCM executor now and it does seem like there's no way to burn and convert a derivative asset back to its reserve asset during the execution of `InitiateReserveWithdraw`, and I think that is the crux of the issue here.

Reply:
@KiChjang 
> What you interact on your local chain should always be `./GeneralIndex(42)` and not `../Parachain(1000)/GeneralIndex(42)` i.e. you interact with the derivative asset and never the reserve asset.

No, I do think we should use reserve asset rather than derivative asset here. except for statemine we need to talk to other parachains like Acala and as you mentioned before only the GeneralIndex is too abstract. How to differ the same derivateive asset `./GeneralIndex(42)` in acala and statemine?

Reply:
Sorry, but your opinion is literally not how XCM works with reserve assets by definition. If you want to use XCM to hold, receive and transfer assets that are not reserve-backed, then you will use asset teleportation instead with your destination chain. Now obviously that would require you to have an agreement with the destination chain to configure each other as trusted teleport locations, and more generally speaking, we will not support a use case where we break the reserve asset model to solve a particular problem.

If you are still inclined to work with this model that you have in mind, then my suggestion would be that you formulate a complete conceptual model, explain why the current reserve asset model does not work and the deficiencies of it, and propose that change in https://github.com/paritytech/xcm-format.

Reply:
@KiChjang Ok, let me put this way, do you think it is possible to support this very basic requirement without introducing new xcm primitives or any hack in cumulus? 
 no big deal for the forward direction from statemine to us, we can do any logic to recognize whatever format defined in statemine. but for the reverse direction, no matter reserved based or teleport based, In current xcm implementation I'am afraid assets will already contains some parachain prefix after the reanchor logic so statemine can not recognize without hack. Correct me if I'am wrong.

Reply:
It seems to me that there's a misunderstanding of what it means when your chain receives a `ReserveAssetDeposited` instruction. `ReserveAssetDeposited` is a _notification_ XCM instruction, meaning it does _not_ actually contain any assets in its parameters; instead, it _notifies_ the recipient that some assets were deposited in their sovereign account on the reserve chain.

Thus, when your chain receives `ReserveAssetDeposited` from the reserve chain, it should then perform the necessary operations to mint a derivative asset that represents the amount of reserve assets that has been transferred to its sovereign account on the reserve chain, and should __not__ be directly minting the reserve asset on your chain, as it is __not__ an asset teleportation.

I suspect this is why you thought it is necessary to use the reserve asset on your chain, and that it is a "very basic" operation. It is not. We do not provide any opinionated approaches as to how you should process and mint derivative assets; it is up to your chain to investigate and invent solutions on how exactly a chain should perform bookkeeping on the assets that they own on the reserve chain. I don't believe this is a new problem however -- there should already been successful examples of how exactly it is done in the ecosystem.

Reply:
@KiChjang I can use the derivative reprensentation(`./GeneralIndex(42)`) in our chain when sent back with `InitiateReserveWithdraw` and actually in a previous [commit](https://github.com/bifrost-finance/bifrost/commit/97476948afda895313ad399048d0ba9471f3fcd9) I did try this way, though which means there are two diffrent repensentations for the same asset when received and sent back which seems weird and actually breaks the principle you mentioned `Assets should always have ONE canonical MultiLocation to represent` I can bear with it.  You may check discussion with @apopiak [here](https://github.com/bifrost-finance/bifrost/pull/420#discussion_r761678588), 

The problem is also what you mentiond  like `there's no way to convert a derivative asset back...during the execution of InitiateReserveWithdraw, and I think that is the crux of the issue here`

Reply:
That could easily be solved if you prepend an `ExchangeAsset` instruction before you execute `InitiateReserveWithdraw`, and in `ExchangeAsset`, your custom XCM executor will convert your derivative asset back to the reserve asset.

This is only one of the ways in which you can solve this; there are many others and I really don't think this is a novel issue -- some other parachains in the ecosystem should have solved this already, and it would be really surprising to me if you're the only chain currently that's running into this problem.

Reply:
> That could easily be solved if you prepend an ExchangeAsset instruction before you execute InitiateReserveWithdraw, and in ExchangeAsset, your custom XCM executor will convert your derivative asset back to the reserve asset.

 @KiChjang I really do not know this primitive and will try, thanks for your suggestion

Reply:
@KiChjang 
seems [ExchangeAsset](https://github.com/paritytech/polkadot/blob/4468927b770eb1515799a9d4aba37b8d1645d787/xcm/xcm-executor/src/lib.rs#L493) you mentioned not implemented yet, is it in some feature branch? And I would like to know if there is any other solution or implementation other parachain team already solved as you mentioned?

Reply:
It is purposefully not implemented; as I said, we do not provide any opinionated approaches as to how a chain should manage their bookkeeping of their reserve assets. I'm not as familiar as to how other parachain teams in the ecosystem solves this problem, however, but I would think that any parachain that has onboarded to Kusama and has dealt with reserve assets would have come up with a solution dealing with it.

Reply:
Understood. we can implemente this to do some asset convert logic from sender side. But I would assume at last we still need `InitiateReserveWithdraw` primitive to send asset back to statemine,right? From what I've tested after the renchor logic in InitiateReserveWithdraw [here](https://github.com/paritytech/polkadot/blob/4468927b770eb1515799a9d4aba37b8d1645d787/xcm/xcm-executor/src/lib.rs#L410) asset will still be converted to `../Parachain(1000)/GeneralIndex(42)` no matter what it previously is and which can not be recognized by statemine which only knows `./GeneralIndex(42)`

Reply:
Correct, the reanchoring logic is what we're in the middle of investigating as well. However, in the context of this PR, your changes will unfortunately not be the correct fix. We will provide an update in #827 as soon as we have any more details to share.